src/zenml/models/v2/core/run_metadata.py

```diff
@@ -13,35 +13,38 @@
 #  permissions and limitations under the License.
 """Models representing run metadata."""
 
-from typing import Dict, Optional
+from typing import Dict, List, Optional
 from uuid import UUID
 
 from pydantic import Field
 
-from zenml.enums import MetadataResourceTypes
 from zenml.metadata.metadata_types import MetadataType, MetadataTypeEnum
 from zenml.models.v2.base.scoped import (
     WorkspaceScopedRequest,
 )
+from zenml.models.v2.misc.run_metadata import RunMetadataResource
 
 # ------------------ Request Model ------------------
 
 
 class RunMetadataRequest(WorkspaceScopedRequest):
     """Request model for run metadata."""
 
-    resource_id: UUID = Field(
-        title="The ID of the resource that this metadata belongs to.",
-    )
-    resource_type: MetadataResourceTypes = Field(
-        title="The type of the resource that this metadata belongs to.",
+    resources: List[RunMetadataResource] = Field(
+        title="The list of resources that this metadata belongs to."
     )
     stack_component_id: Optional[UUID] = Field(
-        title="The ID of the stack component that this metadata belongs to."
+        title="The ID of the stack component that this metadata belongs to.",
+        default=None,
     )
     values: Dict[str, "MetadataType"] = Field(
         title="The metadata to be created.",
     )
     types: Dict[str, "MetadataTypeEnum"] = Field(
         title="The types of the metadata to be created.",
     )
+    cached: Optional[bool] = Field(
+        title="A flag indicating if the run metadata is cached through "
+        "a step execution.",
+        default=False,
+    )
```

Comment:
This is needed to differentiate between the metadata which is linked to a step during its execution with a bare `log_metadata(metadata={})` call and any metadata which is linked to a step manually `log_metadata(metadata={}, step_id="xxx")`. 

Because if a new step is using the cached execution of this step, we want to include metadata coming from the first condition and exclude the metadata related to that second condition.

So if someone runs this pipeline:

```python
from zenml import log_metadata
from zenml import pipeline, step, Model

from zenml.client import Client


@step
def step1() -> None:
    log_metadata(metadata={"score": 0.22})


@step
def step2() -> None:
    log_metadata(metadata={"score": 0.65})
    log_metadata(metadata={"score": 0.55})


@pipeline
def metadata_test():
    step1()
    step2(after="step1")


run = metadata_test()
```

Then calls this:

```python
first_run = client.get_pipeline_run(...)
log_metadata(
    metadata={"score": 0.75, "accuracy": "apt"},
    step_id=first_run.steps["step2"].id
)
```

and then runs the same pipeline again and fetches the metadata of the second run (that is cached):

```python
second_run = client.get_pipeline_run(...)
print(run.run_metadata)
```

They should get: `{'step1::score': 0.22, 'step2::score': 0.55}`

Reply:
I will add a test regarding this use-case as well.

Reply:
I see, in this case, the name looks confusing to me. Maybe `cachable` would do a better job?

Reply:
I have changed the `cached` attribute with `publisher_step_id`. This is for the future where we can use the step execution id to figure out which models got tagged as well.